File: php/Headsort.php
<?php

namespace php;

class Headsort
{
    private array $liste;
    public function __construct(array $liste)
    {
        $this->liste = $liste;
    }

    public function gauche(int $i): int
    {
        return $i * 2 + 1;
    }

    public function droite(int $i): int
    {
        return $i * 2 + 2;
    }

    public function sizeOfListe(): int
    {
        return count($this->liste);
    }

    public function estTas(): bool
    {
        for ($i = 1; $i < ($this->sizeOfListe() - 2) / 2; $i++) {
            if ($this->liste[$this->gauche($i)] > $this->liste[$i]) {
                return false;
            }
            if ($this->droite($i) < count($this->liste) && $this->liste[$this->droite($i)] > $this->liste[$i]) {
                return false;
            }
        }
        return true;
    }

    public function maximun(int $i, int $lim): int
    {
        $maxi = $i;
        $g = $this->gauche($i);
        $d = $this->droite($i);
        if ($g < $lim && $this->liste[$g] > $this->liste[$maxi]) {
            $maxi = $g;
        }
        if ($d < $lim && $this->liste[$d] > $this->liste[$maxi]) {
            $maxi = $d;
        }
        return $maxi;
    }

    public function entasser(int $i, int $lim)
    {
        $maxi = $this->maximun($i, $lim);
        if ($maxi != $i) {
            $temp = $this->liste[$i];
            $this->liste[$i] = $this->liste[$maxi];
            $this->liste[$maxi] = $temp;
            $this->entasser($maxi, $lim);
        }
    }

    public function buildheap()
    {
        $newListe = array_reverse($this->liste);
        for ($i = count($newListe) - 1; $i >= 0; $i--) {
            $this->entasser($i, count($newListe));
        }
    }

    public function trierParTas(): array
    {
        $this->buildheap();
        for ($i = (count($this->liste) - 1); $i > 0; $i--) {
            $temp = $this->liste[0];
            $this->liste[0] = $this->liste[$i];
            $this->liste[$i] = $temp;
            $this->entasser(0, $i);
        }
        return $this->liste;
    }
}

File: haskell/Headsort.hs
module Headsort
  ( Headsort (..),
    trierParTas,
    trierListe,
    maximune,
    entasser,
  )
where

import Control.Concurrent.STM (check)
import Text.XHtml (lime)

newtype Headsort = Headsort {liste :: [Int]}

gauche :: Int -> Int
gauche i = i * 2 + 1

droite :: Int -> Int
droite i = i * 2 + 2

pere :: Int -> Int
pere i = (i - 1) `div` 2

estTas :: Headsort -> Bool
estTas hs = all (\i -> liste hs !! gauche i <= liste hs !! i && (droite i >= length (liste hs) || liste hs !! droite i <= liste hs !! i)) [0 .. ((length (liste hs) - 2) `div` 2)]

maximune :: Headsort -> Int -> Int -> Maybe Int
maximune hs i lim
  | g < lim && liste hs !! g > liste hs !! maxi = Just g
  | d < lim && liste hs !! d > liste hs !! maxi = Just d
  | otherwise = Nothing
  where
    maxi = i
    g = gauche i
    d = droite i

entasser :: Headsort -> Int -> Int -> Headsort
entasser hs i lim
  | maxi /= i = entasser (Headsort new_liste) maxi lim
  | otherwise = hs
  where
    Just maxi = maximune hs i lim
    temp = liste hs !! i
    new_liste = take i (liste hs) ++ [liste hs !! maxi] ++ drop (i + 1) (take len (liste hs)) ++ [temp] ++ drop (lim + 1) (liste hs)
    len = length (liste hs)

buildheap :: Headsort -> Headsort
buildheap hs = foldr (\i acc -> entasser acc i (length (liste hs))) hs [length (liste hs) - 1, length (liste hs) - 2 .. 0]

trierParTas :: Headsort -> [Int]
trierParTas hs = buildheap hs' `seq` foldr (\i acc -> let new_hs = entasser (Headsort (liste hs)) 0 i in new_hs `seq` acc ++ [head (liste new_hs)]) [] [(length (liste hs') - 1), (length (liste hs') - 2) .. 1]
  where
    hs' = Headsort (reverse (liste hs))

trierListe :: [Int] -> [Int]
trierListe xs = trierParTas hs
  where
    hs = Headsort xs